File: src/main/python/AoC2017_11.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 20
#
from __future__ import annotations
from aoc import my_aocd
from aoc.geometry3d import Vector3D, Position3D
import aocd


ORIGIN = Position3D.of(0, 0, 0)
HEADING = {
    "n": Vector3D.of(0, -1, 1),
    "ne": Vector3D.of(1, -1, 0),
    "se": Vector3D.of(1, 0, -1),
    "s": Vector3D.of(0, 1, -1),
    "sw": Vector3D.of(-1, 1, 0),
    "nw": Vector3D.of(-1, 0, 1),
}


def _parse(inputs: tuple[str]) -> list[Vector3D]:
    assert len(inputs) == 1
    return [HEADING[p] for p in inputs[0].split(",")]


def _positions(path: list[Vector3D]) -> list[Position3D]:
    positions = [ORIGIN]
    [positions.append(positions[-1].translate(p)) for p in path]
    return positions


def _steps(p: Position3D) -> int:
    return p.manhattan_distance_to_origin() // 2


def part_1(inputs: tuple[str]) -> int:
    path = _parse(inputs)
    return _steps(_positions(path)[-1])


def part_2(inputs: tuple[str]) -> int:
    path = _parse(inputs)
    return max(_steps(p) for p in _positions(path))


TEST1 = """ne,ne,ne""".splitlines()
TEST2 = """ne,ne,sw,sw""".splitlines()
TEST3 = """ne,ne,s,s""".splitlines()
TEST4 = """se,sw,se,sw,sw""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2017, 11)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST1) == 3
    assert part_1(TEST2) == 0
    assert part_1(TEST3) == 2
    assert part_1(TEST4) == 3

    inputs = my_aocd.get_input(2017, 11, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()

File: src/main/cpp/2017/11/AoC2017_11.cpp
#include <assert.h>

#include <algorithm>
#include <climits>
#include <string>
#include <unordered_map>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aoc/geometry3d/geometry3d.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

const Position3D ORIGIN = Position3D(0, 0, 0);
// clang-format off
const unordered_map<string, Vector3D> HEADING = {
    {"n",  Vector3D(  0, -1,  1)},
    {"ne", Vector3D(  1, -1,  0)},
    {"se", Vector3D(  1,  0, -1)},
    {"s",  Vector3D(  0,  1, -1)},
    {"sw", Vector3D( -1,  1,  0)},
    {"nw", Vector3D( -1,  0,  1)}
};
// clang-format on

vector<Vector3D> parse(const vector<string>& input) {
    assert(input.size() == 1);
    const auto& splits = aoc::split(input[0], ",");
    vector<Vector3D> path;
    transform(splits.begin(), splits.end(), back_inserter(path),
              [](const string s) { return HEADING.at(s); });
    return path;
}

vector<Position3D> positions(const vector<Vector3D>& path) {
    vector<Position3D> positions = {ORIGIN};
    for (const Vector3D& v : path) {
        positions.push_back(positions.back().translate(v));
    }
    return positions;
}

inline int steps(const Position3D& p) {
    return p.manhattanDistance(ORIGIN) / 2;
}

int part1(const vector<string>& input) {
    const vector<Vector3D>& path = parse(input);
    return steps(positions(path).back());
}

int part2(const vector<string>& input) {
    const vector<Vector3D>& path = parse(input);
    int ans = -INT_MAX;
    for (const Position3D& p : positions(path)) {
        ans = max(ans, steps(p));
    }
    return ans;
}

void samples() {
    assert(part1({"ne,ne,ne"}) == 3);
    assert(part1({"ne,ne,sw,sw"}) == 0);
    assert(part1({"ne,ne,s,s"}) == 2);
    assert(part1({"se,sw,se,sw,sw"}) == 3);
}

MAIN(2017, 11)
